File: node/tcptraceroute.js

const PingServer = require('./ping-server');
const TCPClient = require('./tcp-client');
const network = require('network');
const portFinder = require('portfinder');
const dns = require('dns');

function getActiveNetworInterface() {
    return new Promise((resolve, reject) => {
        network.get_active_interface(function(error, active) {
            if (error) {
                reject(error);
            } else {
                resolve(active);
            }
        });
    });
}

function getUnusedPort() {
    return new Promise((resolve, reject) => {
        const minPort = 32 * 1024;
        const maxPort = 64 * 1024
        const randomStart = parseInt(Math.random() * (maxPort - minPort) + minPort);
        portFinder.getPort({port: randomStart, stopPort: maxPort}, function(error, srcPort) {
            if (error) {
                reject(error);
            } else {
                resolve(srcPort);
            }
        });
    });
}

function reverseLookup(ipAddress) {
    return new Promise((resolve, reject) => {
        dns.reverse(ipAddress, function(error, domains) {
            if (error) {
                if (error.code == dns.NOTFOUND) {
                    resolve(ipAddress);
                } else {
                    reject(error);
                }
            } else {
                resolve(domains[0]);
            }
        });
    });
}

function groupByAddress(responses) {
    let byAddress = {}

    for (let idx in responses) {
        if (responses[idx]) {
            if (responses[idx].address in byAddress) {
                byAddress[responses[idx].address].push(responses[idx].time);
            } else {
                byAddress[responses[idx].address] = [responses[idx].time];
            }
        }
    }

    return byAddress;
}

async function showHopResult(hop, icmpResponses, tcpResponses, resolveDomain) {
    let byAddress;

    if (tcpResponses && tcpResponses.length > 0) {
        byAddress = groupByAddress(tcpResponses);
    }

    if (!byAddress) {
        byAddress = groupByAddress(icmpResponses);
    }

    for (let address in byAddress) {
        let domain = address;
        
        if (resolveDomain) {
            domain = await reverseLookup(address);
        }
        console.log(hop.toString().padStart(2, ' '), '', domain, '(' + address + ') ', byAddress[address].join(' ms  '), 'ms');
    }
}

async function trace(dstIp, dstPort, timeout, maxHops, resolveDomain) {
    const activeInterface = await getActiveNetworInterface();
    const srcPort = await getUnusedPort();
    const srcIp = activeInterface.ip_address;

    const client = new TCPClient(srcIp, srcPort, dstIp, dstPort);
    const buffer = client.synPacket();

    console.log(`Selected device ${activeInterface.name}, address ${srcIp}, port ${srcPort} for outgoing packets`);
    console.log(`Tracing the path to ${dstIp} on TCP port ${dstPort}, ${maxHops} hops max`);

    for (let hop = 1; hop <= maxHops; hop = hop + 1) {
        let icmpResponses = [];
        let tcpResponses = [];
        let icmpResponse;

        for (let trip = 0; trip < 3; trip = trip + 1) {
            icmpResponse = await PingServer.ping(dstIp, hop, 5000);
            icmpResponses.push(icmpResponse);
         
            if (!icmpResponse || icmpResponse.address !== dstIp) {
                tcpResponses.push(await PingServer.ping(dstIp, hop, timeout, client, buffer));
            }
        }

        await showHopResult(hop, icmpResponses, tcpResponses, resolveDomain);

        if (icmpResponse && icmpResponse.address === dstIp) {
            break;
        }
    }

    client.close();
}

const args = process.argv.slice(2);

if (args.length < 2) {
    console.log('Usage: node tcptraceroute.js host port');
} else {
    try {
        trace(args[0], parseInt(args[1]), 5000, 30, true)
            .catch((error) => {
                if (error.toString().includes('Operation not permitted')) {
                    console.log('Got root?');
                } else {
                    console.log(error);
                }
            });
    }
    catch (error) {
    }
}

File: python/tcptraceroute.py

import sys
import socket
import psutil
from pingserver import PingServer
from tcpclient import TCPClient


def getActiveNetworInterface(dstIp, dstPort):
    s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
    s.connect((dstIp, dstPort))
    interfaceIp = s.getsockname()[0]

    interfaces = psutil.net_if_addrs()
    interfaceName = [i for i in interfaces for j in interfaces[i] if j.address==interfaceIp and j.family==socket.AF_INET][0]

    return interfaceName, interfaceIp


def groupByAddress(responses):
    byAddress = {}

    for response in responses:
        if response:
            if response['address'] in byAddress:
                byAddress[response['address']].append(response['time'])
            else:
                byAddress[response['address']] = [response['time']]

    return byAddress


def reverseLookup(address):
    try:
        name, alias, addressList = socket.gethostbyaddr(address)
        return name
    except socket.herror:
        return address


def showHopResult(hop, icmpResponses, tcpResponses, resolveDomain):
    byAddress = {}

    if tcpResponses:
        byAddress = groupByAddress(tcpResponses)

    if not byAddress:
        byAddress = groupByAddress(icmpResponses)

    for address in byAddress:
        domain = address;
        
        if resolveDomain:
            domain = reverseLookup(address)

        print('{:2d}  {} ({})  {} ms'.format(hop, domain, address, ' ms  '.join(map('{0:.2f}'.format, byAddress[address]))))


def trace(dstIp, dstPort, timeout, maxHops, resolveDomain):
    interfaceName, srcIp = getActiveNetworInterface(dstIp, dstPort)
    srcPort = 31234

    client = TCPClient(srcIp, srcPort, dstIp, dstPort)
    buffer = client.synPacket()

    print('Selected device {}, address {}, port {:d} for outgoing packets'.format(interfaceName, srcIp, srcPort))
    print('Tracing the path to {} on TCP port {:d}, ${:d} hops max'.format(dstIp, dstPort, maxHops))

    for hop in range(1, maxHops + 1):
        icmpResponses = []
        tcpResponses = []
        icmpResponse = None

        for trip in range(0, 3):
            icmpResponse = PingServer.ping(dstIp, hop, timeout)
            icmpResponses.append(icmpResponse)
         
            if not icmpResponse or icmpResponse['address'] != dstIp:
                tcpResponses.append(PingServer.ping(dstIp, hop, timeout, client, buffer))

        showHopResult(hop, icmpResponses, tcpResponses, resolveDomain)

        if icmpResponse and icmpResponse['address'] == dstIp:
            break


args = sys.argv

if len(args) < 3:
    print('Usage: python tcptraceroute.py host port');
else:
    try:
        trace(args[1], int(args[2]), 5000, 30, False)
    except PermissionError:
        print('Got root?')
    except Exception as ex:
        print(ex)

# trace('8.8.8.8', 443, 5000, 30, True)
